<h1 style = "text-align: center;"> Peer-graded Assignment </h1>
<h3 style = "text-align: center;"> Segmenting and Clustering Neighborhoods in Toronto </h3>

<h5>Get Libraries and Packages</h5>

In [1]:
!pip install bs4
import pandas as pd
import requests
from bs4 import BeautifulSoup

<h5>Dataframe from Part 1</h5>

In [2]:
#Read data from html
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
html_data = requests.get(url).text

#Parse data
soup = BeautifulSoup(html_data, 'lxml') 

#Sort data
table_contents=[]
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)

# print(table_contents)
df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

df.head()

PostalCode           Borough                      Neighborhood
0        M3A        North York                         Parkwoods
1        M4A        North York                  Victoria Village
2        M5A  Downtown Toronto         Regent Park, Harbourfront
3        M6A        North York  Lawrence Manor, Lawrence Heights
4        M7A      Queen's Park     Ontario Provincial Government

<h2 style = "text-align: center;"> Part 2: Getting Location Data of Each Neighborhood </h2>

<h3>Create the Postal Code DataFrame</h3>

In [3]:
df_postcode = df.groupby(['PostalCode', 'Borough'])['Neighborhood'].apply(', '.join).reset_index()
df_postcode.head()

PostalCode      Borough                            Neighborhood
0        M1B  Scarborough                          Malvern, Rouge
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2        M1E  Scarborough       Guildwood, Morningside, West Hill
3        M1G  Scarborough                                  Woburn
4        M1H  Scarborough                               Cedarbrae

<h3>Read .csv file with Geospatial Data </h3>

In [4]:
geo = pd.read_csv('https://cocl.us/Geospatial_data')
geo.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

<h6>Rename Columns</h6>

In [5]:
geo.columns = ['PostalCode', 'Latitude', 'Longitude']
geo.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

<h3>Merge the DataFrames</h3>

In [6]:
df_combo = pd.merge(df_postcode, geo, on = 'PostalCode')
df_combo.head()

PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Malvern, Rouge  43.806686   
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

<h5>Check Successful Merging</h5>

In [7]:
df_combo[df_combo["PostalCode"] == "M5G"]

PostalCode           Borough        Neighborhood   Latitude  Longitude
57        M5G  Downtown Toronto  Central Bay Street  43.657952 -79.387383

Confirmed that the postal code M5G is as it should be which is at (43.657952, -79.387383)